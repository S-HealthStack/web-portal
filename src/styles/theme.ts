import { css, DefaultTheme } from 'styled-components';
import { px } from './utils';

const colors = {
  // gray gradations
  black: '#000000',
  black40: 'rgba(0, 0, 0, 0.4)',
  black08: 'rgba(0, 0, 0, 0.08)',

  // light mode
  surface: '#FFFFFF',
  background: '#F7F8FA',
  onPrimary: '#FFFFFF',
  backgroundOnPrimary: '#FFFFFF',
  backgroundSurface: '#FFFFFF',
  backgroundOnSurface: '#2D2D2D',
  primary: '#4475E3',
  primaryLight: '#ECF1FC',
  primaryHovered: '#2F65E0',
  primary05: '#F6F8FE',
  primary10: '#ECF1FC',
  primary30: '#C7D6F7',
  primaryDisabled: '#B3C6F1',
  primaryLightPressed: '#E3EAFB',
  primaryLightFocused: '#C7D6F7',
  textPrimary: '#474747',
  primaryWhite: '#FFFFFF',
  disabled: '#DBDBDB',
  disabled20: 'rgba(219, 219, 219, 0.2)',
  onDisabled: '#A1A1A1',

  textPrimaryBlue: '#4475E3',
  textPrimaryDark: '#474747',
  textSecondaryGray: '#808080',
  textDisabled: '#838383',
  onSurface: '#2D2D2D',
  secondarySkyblue: '#00B0D7', // replace
  backgoundLight: '#F0F0F0',
  primaryBlueHovered: '#2F65E0',
  primaryBluePressed: '#2057D5',

  statusSuccess: '#2DC008',
  statusSuccess10: '#DCF2EA',
  statusSuccessText: '#00825D',

  statusError: '#FF3333',
  statusError10: '#FFEBEB',
  statusErrorText: '#D14343',

  statusWarning: '#F5BF00',
  statusWarning10: '#FFF7D0',
  statusWarningText: '#A36400',

  secondaryViolet: '#944ED7',
  secondaryVioletHovered: '#6A3799',
  secondaryViolet10: '#F4EDFB',

  secondaryTangerine: '#F39C18',
  secondaryTangerineHovered: '#CC8314',
  secondaryTangerine10: '#FEF5E8',

  secondarySkyBlue: '#00B0D7',
  secondarySkyBlueHovered: '#0092B3',
  secondarySkyBlue10: '#E5F7FB',

  secondaryGreen: '#1AD598',
  secondaryGreenHovered: '#16B37F',
  secondaryGreen10: '#E8FBF5',

  secondaryRed: '#FA5F4F',
  secondaryRedHovered: '#CC4E40',
  secondaryRed10: '#FFEFED',
};

export type SpecColorType = keyof typeof colors;

const weightToCssValue = {
  regular: 400,
  upperregular: 450,
  medium: 500,
  semibold: 600,
  bold: 700,
};

const fontSpec = (
  size: number,
  lineHeight: number | string,
  weight: keyof typeof weightToCssValue,
  fontFamily: string
) => css`
  font-family: ${fontFamily}, sans-serif;
  font-size: ${px(size)};
  line-height: ${typeof lineHeight === 'string' ? lineHeight : px(lineHeight)};
  font-weight: ${weightToCssValue[weight]};
  font-style: normal;
`;

const typography = {
  // Inter typeface
  labelRegular: fontSpec(10, 15, 'regular', 'Inter'),
  labelSemibold: fontSpec(10, 15, 'semibold', 'Inter'),

  bodyXSmallRegular: fontSpec(12, 18, 'regular', 'Inter'),
  bodyXSmallSemibold: fontSpec(12, 18, 'semibold', 'Inter'),
  bodySmallRegular: fontSpec(14, 21, 'regular', 'Inter'),
  bodySmallSemibold: fontSpec(14, 21, 'semibold', 'Inter'),
  bodyMediumRegular: fontSpec(16, 24, 'regular', 'Inter'),
  bodyMediumSemibold: fontSpec(16, 24, 'semibold', 'Inter'),
  bodyLargeRegular: fontSpec(18, 26, 'regular', 'Inter'),
  bodyLargeSemibold: fontSpec(18, 26, 'semibold', 'Inter'),

  headingXSmall: fontSpec(16, 21, 'semibold', 'Inter'),
  headingSmall: fontSpec(18, 23, 'semibold', 'Inter'),
  headingXMedium: fontSpec(18, 27, 'semibold', 'Inter'),
  headingXMediumRegular: fontSpec(18, 27, 'regular', 'Inter'),
  headingMedium: fontSpec(24, 32.2, 'semibold', 'Inter'),
  headingLargeSemibold: fontSpec(32, 42, 'semibold', 'Inter'),
  headingXLargeSemibold: fontSpec(48, 58, 'semibold', 'Inter'),
  headingXXLargeSemibold: fontSpec(60, 72, 'semibold', 'Inter'),

  // Fira Code typeface
  query14: fontSpec(14, '130%', 'upperregular', 'Fira Code'),
  query12: fontSpec(12, '150%', 'upperregular', 'Fira Code'),
};

const boxShadow = {
  card: `0 0 ${px(2)} ${px(0)} rgba(0,0,0,0.15)`,
  avatar: `0 ${px(8)} ${px(10)} ${px(2)} 0 rgba(0,0,0,0.12)`,
  studyAvatar: `0 ${px(15)} ${px(20)} rgba(180,180,180,0.3)`,
  previewScreen: `0 ${px(2)} ${px(4)} rgba(71,71,71,0.25)`,
};

const animation = {
  defaultTiming: 'cubic-bezier(0.45, 0.05, 0, 1)',
  rippleTiming: 'cubic-bezier(0.4, 0, 0.2, 1)',
};

declare module 'styled-components' {
  export interface DefaultTheme {
    colors: typeof colors;
    typography: typeof typography;
    boxShadow: typeof boxShadow;
    animation: typeof animation;
  }
}

const theme: DefaultTheme = {
  colors,
  typography,
  boxShadow,
  animation,
};

export default theme;
