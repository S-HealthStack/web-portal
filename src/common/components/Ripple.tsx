import React, { useCallback, useEffect, useMemo, useRef, useState } from 'react';
import styled, { css, keyframes } from 'styled-components';
import _uniqueId from 'lodash/uniqueId';
import _isEqual from 'lodash/isEqual';

import { px } from 'src/styles';
import { SpecColorType } from 'src/styles/theme';
import combineRefs from 'src/common/utils/combineRefs';

export const DEFAULT_RIPPLE_ANIMATION_DURATION = 600;

export interface RipplePos {
  left: number;
  top: number;
}

export interface RippleProps extends React.HTMLAttributes<HTMLDivElement> {
  ref: React.RefObject<HTMLDivElement>;
  key: React.Key;
  isRippling: boolean;
  size: number;
  position: RipplePos;
  color?: SpecColorType;
  opacity: [number, number];
  duration?: number;
  scale: number;
  isEntered: boolean;
  isExited: boolean;
}

export interface UseRippleArgs {
  duration?: number;
  opacity?: number | [number, number];
  color?: SpecColorType;
}

export interface RippleTriggerProps<T> extends React.PropsWithChildren<object> {
  ref: React.RefObject<T> | ((commonRef: T | null) => void);
  style: Pick<React.CSSProperties, 'position' | 'overflow'>;
  onMouseDown: (evt: React.MouseEvent<T>) => void;
  onMouseUp: (evt: React.MouseEvent<T>) => void;
  onMouseLeave: (evt: React.MouseEvent<T>) => void;
}

type HandleRippleInOpts = { force?: boolean };
type HandleRippleOutOpts = { force?: boolean };

export interface UseRippleReturn<
  T,
  P extends Partial<RippleTriggerProps<T>> | undefined = undefined
> {
  rippleTriggerProps: RippleTriggerProps<T>;
  addRippleTriggerProps: (props?: P) => P;
  rippleProps: RippleProps;
  handleRippleIn: (evt?: React.MouseEvent<T>, opts?: HandleRippleInOpts) => void;
  handleRippleOut: (evt?: React.MouseEvent<T>, opts?: HandleRippleOutOpts) => void;
}

const rippleOpacity = (opacity?: [number, number]): [number, number] =>
  typeof opacity !== 'undefined' ? opacity : [1, 1];

const rippleInAnimation = (opacity: [number, number], scale: number) => keyframes`
  0% {
    transform: scale(0);
    opacity: ${rippleOpacity(opacity)[0]};
  }
  100% {
    transform: scale(${scale});
    opacity: ${rippleOpacity(opacity)[1]};
  }
`;

const rippleOutAnimation = (opacity: [number, number], scale: number) => keyframes`
  0% {
    opacity: ${rippleOpacity(opacity)[1]};
    transform: scale(${scale});
  }
  100% { 
    opacity: 0;
    transform: scale(${scale});
  }
`;

const rippleDuration = (duration?: number) =>
  typeof duration !== 'undefined' ? duration : DEFAULT_RIPPLE_ANIMATION_DURATION;

// TODO: add the ability to fire multiple ripples
export const useRipple = <T extends HTMLElement, P extends Partial<RippleTriggerProps<T>> = object>(
  args: UseRippleArgs
): UseRippleReturn<T, P> => {
  const rippleRef = useRef<HTMLDivElement>(null);
  const targetRef = useRef<T>(null);

  const [isMouseHold, setMouseHold] = useState<boolean>(false);
  const [isEnterAnimationEnd, setEnterAnimationEnd] = useState<boolean>(false);
  const [isExitAnimationEnd, setExitAnimationEnd] = useState<boolean>(false);
  const [isRippling, setRippling] = useState<boolean>(false);
  const [position, setPosition] = useState<RipplePos>({ left: 0, top: 0 });
  const [scale, setScale] = useState<number>(0);
  const [size, setSize] = useState<number>(0);
  const [key, setKey] = useState<string>('');
  const [duration, setDuration] = useState<number>(rippleDuration(args.duration));

  const handleRippleAnimationEnd = useCallback(() => {
    !isEnterAnimationEnd ? setEnterAnimationEnd(true) : setExitAnimationEnd(true);
  }, [isEnterAnimationEnd]);

  const handleRippleIn = useCallback(
    (evt?: React.MouseEvent<T>, opts?: HandleRippleInOpts) => {
      const currentTarget = evt ? (evt.currentTarget as T) : targetRef.current;

      if (!currentTarget) {
        throw new Error(
          `You are probably using ripple without 'React.MouseEvent'.` +
            `Set the 'ref' returned from 'useRipple().rippleTriggerProps.ref' to the target component.`
        );
      }

      const elementRect = currentTarget.getBoundingClientRect();

      // set relative position of cursor in element or calculate averages if evt doesn't exist
      const cursorRelativePositionX = evt
        ? evt.pageX - elementRect.x
        : currentTarget.clientWidth / 2;
      const cursorRelativePositionY = evt
        ? evt.pageY - elementRect.y
        : currentTarget.clientHeight / 2;

      // length of segments
      const leftX = cursorRelativePositionX;
      const rightX = currentTarget.clientWidth - cursorRelativePositionX;
      const topY = cursorRelativePositionY;
      const bottomY = currentTarget.clientHeight - cursorRelativePositionY;

      // circle radius: sqrt(pow(x2 - x1) + pow(y2 - y1))
      const circleRadius = Math.sqrt(
        (currentTarget.clientWidth - Math.min(leftX, rightX)) ** 2 +
          (currentTarget.clientHeight - Math.min(topY, bottomY)) ** 2
      );

      // max scale and size
      const newSize = Math.min(currentTarget.clientHeight, currentTarget.clientWidth);
      const maxScale = (circleRadius / newSize) * 2;

      setEnterAnimationEnd(false);
      setExitAnimationEnd(false);
      setMouseHold(true);
      setScale(maxScale);
      setSize(newSize);
      setKey(_uniqueId());
      setRippling(true);
      setDuration(opts?.force ? 0 : rippleDuration(args.duration));
      setPosition({
        left: cursorRelativePositionX - newSize / 2,
        top: cursorRelativePositionY - newSize / 2,
      });
    },
    [args.duration]
  );

  const handleRippleOut = useCallback((_?: React.MouseEvent<T>, opts?: HandleRippleOutOpts) => {
    setMouseHold(false);
    if (opts?.force) {
      setEnterAnimationEnd(false);
      setExitAnimationEnd(false);
      setRippling(false);
    }
  }, []);

  useEffect(() => {
    if (!isMouseHold && isEnterAnimationEnd && isExitAnimationEnd) {
      setRippling(false);
    }
  }, [isMouseHold, isEnterAnimationEnd, isExitAnimationEnd]);

  const opacity: [number, number] = useMemo(() => {
    if (typeof args.opacity !== 'undefined') {
      return Array.isArray(args.opacity) ? args.opacity : [args.opacity, args.opacity];
    }
    return [1, 1];
  }, [args.opacity]);

  const rippleTriggerProps: RippleTriggerProps<T> = {
    ref: targetRef,
    onMouseDown: handleRippleIn,
    onMouseUp: handleRippleOut,
    onMouseLeave: handleRippleOut,
    style: {
      overflow: 'hidden',
      position: 'relative',
    },
  };

  const addRippleTriggerProps = useCallback(
    (props?: P): P => {
      const newProps: P = (props ? { ...props } : {}) as P;

      if (props?.ref) {
        newProps.ref = combineRefs<T>([rippleTriggerProps.ref, props.ref]);
      } else {
        newProps.ref = targetRef;
      }

      newProps.onMouseDown = (evt) => {
        handleRippleIn(evt);
        props?.onMouseDown?.(evt);
      };

      newProps.onMouseUp = (evt) => {
        handleRippleOut();
        props?.onMouseUp?.(evt);
      };

      newProps.onMouseLeave = (evt) => {
        handleRippleOut();
        props?.onMouseLeave?.(evt);
      };

      if (props?.style) {
        newProps.style = {
          ...newProps?.style,
          ...rippleTriggerProps.style,
        };
      } else {
        newProps.style = rippleTriggerProps.style;
      }

      return newProps;
    },
    [handleRippleIn, handleRippleOut, rippleTriggerProps.ref, rippleTriggerProps.style]
  );

  const rippleProps: RippleProps = {
    ref: rippleRef,
    key,
    isRippling,
    size,
    position,
    duration,
    scale,
    color: args.color,
    opacity,
    onAnimationEnd: handleRippleAnimationEnd,
    isEntered: !isEnterAnimationEnd || isMouseHold,
    isExited: !isExitAnimationEnd,
  };

  return {
    handleRippleIn,
    handleRippleOut,
    rippleTriggerProps,
    addRippleTriggerProps,
    rippleProps,
  };
};

const getRippleStyles = ({
  size,
  position,
  isRippling,
}: RippleProps): Partial<CSSStyleDeclaration> => ({
  display: isRippling ? 'block' : 'none',
  width: px(size),
  height: px(size),
  top: px(position.top),
  left: px(position.left),
});

const Ripple = styled.div
  .withConfig({
    shouldForwardProp: (prop, defaultValidatorFn) =>
      defaultValidatorFn(prop) && !['in'].includes(prop),
  })
  .attrs<RippleProps>((props) => ({
    style: getRippleStyles(props),
  }))<RippleProps>`
  position: absolute;
  will-change: opacity, transform;
  border-radius: 50%;
  animation: ${({ isEntered, opacity, scale, duration, theme }) =>
    isEntered
      ? css`
          ${rippleInAnimation(opacity, scale)} ${rippleDuration(duration)}ms ${theme.animation
            .rippleTiming} forwards
        `
      : css`
          ${rippleOutAnimation(opacity, scale)} 300ms ${theme.animation.defaultTiming} forwards
        `};
  z-index: 0;
  background-color: ${({ color, theme }) => theme.colors[color || 'primaryLight']};
  pointer-events: none;
`;

export default React.memo(Ripple, _isEqual);
