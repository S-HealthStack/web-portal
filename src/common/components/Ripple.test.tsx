import React from 'react';
import '@testing-library/jest-dom';
import 'jest-styled-components';
import { render } from '@testing-library/react';
import theme from 'src/styles/theme';
import { ThemeProvider } from 'styled-components';
import { RippleTest } from './Ripple.stories';

describe('Ripple', () => {
  it('test ripple render', async () => {
    const { baseElement, queryByTestId } = render(
      <ThemeProvider theme={theme}>
        <RippleTest />
      </ThemeProvider>
    );

    expect(baseElement).toMatchSnapshot();

    const ripple = queryByTestId('ripple') as Element;
    const trigger = queryByTestId('trigger') as Element;

    expect(ripple).not.toBeVisible();
    expect(trigger).toBeVisible();
  });
});
