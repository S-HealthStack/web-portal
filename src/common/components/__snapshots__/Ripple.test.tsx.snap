// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Ripple test ripple render 1`] = `
.c1 {
  position: absolute;
  will-change: opacity,transform;
  border-radius: 50%;
  -webkit-animation: bykygw 600ms cubic-bezier(0.4,0,0.2,1) forwards;
  animation: bykygw 600ms cubic-bezier(0.4,0,0.2,1) forwards;
  z-index: 0;
  background-color: #ECF1FC;
  pointer-events: none;
}

.c0 {
  width: 16rem;
  height: 16rem;
  border: 1px solid lightblue;
}

<body>
  <div>
    <div
      class="c0"
      data-testid="trigger"
      style="overflow: hidden; position: relative;"
    >
      <div
        class="c1"
        data-testid="ripple"
        opacity="1,1"
        scale="0"
        style="display: none; width: 0rem; height: 0rem; top: 0rem; left: 0rem;"
      />
    </div>
  </div>
</body>
`;
